### PR TITLE
Refact notifications

### PR DESCRIPTION
**Pull request by Luiz-FS**
**Feature/Bug description:** The current architecture of the notifications was causing much confusion during the development of new notifications and the refactoring of those that already exist. Currently, in the system to create a new notification, it is necessary to pass to the send_notification method the type of the notification, the user who did the action, the current user's institution and the key of the entity that will be retrieved in the frontend when the user clicks in the notification. 
After the send_notification method is called, it creates a notification message in the following format:
```javascript
Message = {
    'from': {
        'name': ...,  /*name of the user who made the action.*/
        'photo_url': ..., /*photo_url of the user who made the action.*/
        'insitution_name': ..., /*current institution name of user.*/
    }
}

Entity = {
    'key': ..., /*key of the object to be retrieved.*/
    'institution_name': .../*Name of the institution associated with the past object.*/
}
```

This architecture caused a lot of confusion and inconsistency in the development of the notifications, especially when there were two institutions involved that needed to be in the notification because the name of the institution was always being taken by the object that was passed by parameter and this could not be changed dynamically.

**Solution:** We reformulated the architecture of the notifications in order to prevent inconsistencies and facilitate the understanding of each attribute of the notification and also allow the institutions to be changed in the message according to what is needed in each notification.
  Now models that create their own standard messages and send them to the send_notification method. In this way it was possible to create more dynamic messages and improve their understanding. In the current file, we have separated two new attributes, 'current_institution' and 'to', to separate the message correctly. The attribute 'current_insitution' contains the name of the institution in which the user made the action, 'from' contains the name of the institution for which the notification was created and the 'to' contains the institution in which the notification was directed (In the case where two institutions are required).
The new notifications will follow the following format:

```javascript
Message = {
    'current_institution': {
        'name': ... /*current institution name of user.*/
    },
    'from': {
        'name': ...,  /*name of the user who made the action.*/
        'photo_url': ..., /*photo_url of the user who made the action.*/
        'insitution_name': ..., /*name of the institution for which the notification was created.*/
    },
    'to': {
        'institution_name': ... /*institution in which the notification was directed (In the case where two institutions are required).*/
    }
}

Entity = {
    'key': ... /*key of the object to be retrieved.*/
}
```

**TODO/FIXME:** n/a

### CODE REVIEW

File: backend/handlers/institution_parent_request_handler.py [hunk 1]
```diff
@@ -55,4 +55,5 @@ def delete(self, user, request_key):
         request.change_status('rejected')
         request.put()
 
-        request.send_response_notification(request.institution_requested_key, user.key, 'REJECT')
+        # PASSAR LOGO O TIPO DA NOTIF
```
Comment (raonismaneoto) [inline backend/handlers/institution_parent_request_handler.py]: remove it

File: backend/models/institution.py [hunk 2]
```diff
@@ -383,6 +384,22 @@ def get_super_user_admin_permissions(self, permissions=None):
             
         return permissions
 
+    def create_notification_message(self, user_key, current_institution_key, 
+            receiver_institution_key=None, sender_institution_key=None):
+        """ Create message that will be used in notification. 
+            user_key -- The user key that made the action.
+            current_institution -- The institution that user was in the moment that made the action.
+            sender_institution_key -- The institution that should be made the action,
```
Comment (raonismaneoto) [inline backend/models/institution.py]: This description is confusing

File: backend/models/invite.py [hunk 3]
```diff
@@ -85,22 +86,37 @@ def send_email(self, host, receiver_email=None, body=None):
         })
         email_sender.send_email()
 
-    def send_notification(self, current_institution, sender_key=None, receiver_key=None, entity_type=None, entity_key=None):
+    def create_notification_message(self, user_key, current_institution_key=None, 
+            sender_institution_key=None, receiver_institution_key=None):
+        """ Create message that will be used in notification. 
+            user_key -- The user key that made the action.
+            current_institution -- The institution that user was in the moment that made the action,
```
Comment (raonismaneoto) [inline backend/models/invite.py]: current_institution_key*

File: backend/models/invite.py [hunk 4]
```diff
@@ -85,22 +86,37 @@ def send_email(self, host, receiver_email=None, body=None):
...
+                 in case that user is inactive he didn't have institution.
+            sender_institution_key -- The institution that should be made the action.
```
Comment (raonismaneoto) [inline backend/models/invite.py]: again, confusing

File: backend/models/post.py [hunk 5]
```diff
@@ -347,6 +348,20 @@ def remove_subscriber(self, user):
         if user.key in self.subscribers and self.author != user.key:
             self.subscribers.remove(user.key)
 
+    def create_notification_message(self, user_key, current_institution_key, sender_institution_key=None):
+        """ Create message that will be used in notification.
+            user_key -- The user key that made the action.
+            current_institution -- The institution that user was in the moment that made the action.
```
Comment (raonismaneoto) [inline backend/models/post.py]: _key*

File: backend/models/post.py [hunk 6]
```diff
@@ -347,6 +348,20 @@ def remove_subscriber(self, user):
...
+            sender_institution_key -- The institution by which the post was created,
+                if didn't define the sender institution  should be the current institution.
```
Comment (raonismaneoto) [inline backend/models/post.py]: if it hasn't been defined yet, the sender institution should be the current institution..

File: backend/models/request.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 """Request Model."""
 from models.invite import Invite
 from custom_exceptions.fieldException import FieldException
+from service_messages import create_message
```
Comment (raonismaneoto) [inline backend/models/request.py]: Is it necessary?

File: backend/models/request_user.py [hunk 8]
```diff
@@ -94,27 +95,37 @@ def send_response_email(self, host, operation):
...
+            entity_key=self.key.urlsafe(),
+            message=notification_message
         )
 
     """ 
     The sender, in this case, is the user who is asking to be an institution's members.
     The user is who answered the request.
     The institution_key to which the sender requested to be member.
-    Entity_type can be REJECTED_LINK or ACCEPTED_LINK.
+    Action can be REJECTED_LINK or ACCEPTED_LINK.
```
Comment (raonismaneoto) [inline backend/models/request_user.py]: REJECTED_LINK or just REJECTED? same to accept.

File: backend/service_messages.py [hunk 9]
```diff
@@ -11,63 +11,53 @@
...
 sys.setdefaultencoding('utf8')
 
-def create_message(sender_key, current_institution):
-    """Create a message based on user current institution."""
-    sender = ndb.Key(urlsafe=sender_key).get()
+def create_message(sender_key, current_institution_key=None, receiver_institution_key=None, sender_institution_key=None):
+    """Create a message of notification.
+    
+        Keyword arguments:
+        sender_key -- the user key of who is sending the notification
```
Comment (raonismaneoto) [inline backend/service_messages.py]: the key of the user who is sending the notification

File: backend/service_messages.py [hunk 10]
```diff
@@ -11,63 +11,53 @@
...
-        institution_name = current_institution.get().name 
-    else:
-        institution_name = ""
+
     message = {
         'from': {
             'name': name.encode('utf8'),
             'photo_url': sender.photo_url,
-            'institution_name': institution_name
+            'institution_name': (sender_institution_key and sender_institution_key.get().name) or ''
```
Comment (raonismaneoto) [inline backend/service_messages.py]: this expression can return True or False, can't it?

File: backend/service_messages.py [hunk 11]
```diff
@@ -11,63 +11,53 @@
...
+        },
+        'to': {
+            'institution_name': (receiver_institution_key and receiver_institution_key.get().name) or ''
```
Comment (raonismaneoto) [inline backend/service_messages.py]: same thing

File: backend/service_messages.py [hunk 12]
```diff
@@ -11,63 +11,53 @@
...
+        },
+        'current_institution': {
+            'name': current_institution_key and current_institution_key.get().name
```
Comment (raonismaneoto) [inline backend/service_messages.py]: same thing

File: backend/models/institution.py (hunk 2 above)
Comment (ruanres) [inline backend/models/institution.py]: The institution in which action should be taken
Acho que assim fica melhor

File: backend/models/institution.py [hunk 13]
```diff
@@ -383,6 +384,22 @@ def get_super_user_admin_permissions(self, permissions=None):
...
+                 when wasn't specified will be the current_institution.
```
Comment (ruanres) [inline backend/models/institution.py]: when it is not specified it will...

File: frontend/notification/notificationService.js [hunk 14]
```diff
@@ -54,8 +53,8 @@
 
         service.formatMessage = function formatMessage(notification) {
             var entity_type = notification.entity_type;
-            var mainInst = notification.entity.institution_name;
-            var otherInst = notification.from.institution_name;
+            var mainInst = notification.from.institution_name;
```
Comment (mayzabeel) [inline frontend/notification/notificationService.js]: Think if possible, show the new notifications and the old ones. Maybe check if notification hás a new param to decide what show

File: backend/models/invite.py [hunk 15]
```diff
@@ -85,22 +86,37 @@ def send_email(self, host, receiver_email=None, body=None):
         })
         email_sender.send_email()
 
-    def send_notification(self, current_institution, sender_key=None, receiver_key=None, entity_type=None, entity_key=None):
+    def create_notification_message(self, user_key, current_institution_key=None, 
+            sender_institution_key=None, receiver_institution_key=None):
+        """ Create message that will be used in notification. 
+            user_key -- The user key that made the action.
+            current_institution -- The institution that user was in the moment that made the action,
+                 in case that user is inactive he didn't have institution.
```
Comment (ruanres) [inline backend/models/invite.py]: melhorar

File: backend/models/invite.py (hunk 4 above)
Comment (ruanres) [inline backend/models/invite.py]: melhorar

File: backend/models/request_user.py (hunk 8 above)
Comment (ruanres) [inline backend/models/request_user.py]: Action can be ACCEPT or REJECTED

File: backend/service_messages.py [hunk 16]
```diff
@@ -11,63 +11,53 @@
...
-    if is_post:
-        institution = entity_obj.institution.get()
-        name = institution.name
-    elif is_institution:
-        name = entity_obj.name
-    elif is_invite or is_request:
-        institution = entity_obj.institution_key.get()
-        name = institution.name
-    
+    """Create a short entity with only key and name."""
```
Comment (ruanres) [inline backend/service_messages.py]: essa entity so tem a prop key

File: backend/worker.py [hunk 17]
```diff
@@ -307,22 +321,23 @@ def post(self):
...
         entity_type = self.request.get('entity_type')
         entity_key = self.request.get('entity_key')
-        current_institution = ndb.Key(urlsafe=self.request.get('current_institution'))
+        current_institution_key = ndb.Key(urlsafe=self.request.get('current_institution'))
         entity = self.request.get('entity') if self.request.get('entity') else None
-        
         inst_key = self.request.get('institution_key')
         institution = ndb.Key(urlsafe=inst_key).get()
+        
+        obj =  ndb.Key(urlsafe=entity_key).get() if(entity_key) else institution
```
Comment (ruanres) [inline backend/worker.py]: remover espaço

File: backend/models/invite.py (hunk 4 above)
Comment (ruanres) [reply backend/models/invite.py]: Melhorar nos outros modelos tbm

File: backend/handlers/institution_members_handler.py [hunk 18]
```diff
@@ -39,8 +39,8 @@ def delete(self, user, url_string):
                               url_string)
 
         institution = institution_key.get()
-        data = self.request.get('removeMember')
-        member = ndb.Key(urlsafe=data)
+        member_key = self.request.get('removeMember')
```
Comment (raonismaneoto) [inline backend/handlers/institution_members_handler.py]: shouldn't this variable be named as memberToRemove?

File: frontend/notification/notificationService.js [hunk 19]
```diff
@@ -53,8 +53,8 @@
 
         service.formatMessage = function formatMessage(notification) {
             var entity_type = notification.entity_type;
-            var mainInst = notification.from.institution_name;
-            var otherInst = notification.to.institution_name;
+            var mainInst = notification.entity.institution_name || notification.from.institution_name;
```
Comment (mayzabeel) [inline frontend/notification/notificationService.js]: insert a comment explaining what is new notification data and what is old notification data

File: frontend/notification/notifications.html [hunk 20]
```diff
@@ -17,7 +17,7 @@
             <img ng-src="{{notification.from.photo_url}}" class="md-avatar" style="width: 60px; height: 60px;"/>
             <div class="md-list-item-text" layout="column" >
               <p><b style="font-size: 18px;">{{ notificationCtrl.limitString(notification.from.name, 28) }}</b></p>
-              <p><b style="font-size: 15px; color: #009688;">{{ notificationCtrl.limitString(notification.current_institution.name, 32) }}</b></p>
+              <p><b style="font-size: 15px; color: #009688;">{{ notificationCtrl.limitString(notification.current_institution.name, 32) || notificationCtrl.limitString(notification.from.institution_name, 32)}}</b></p>
```
Comment (mayzabeel) [inline frontend/notification/notifications.html]: insert a comment explaining what is new notification data and what is old notification data

File: frontend/notification/notifications_page.html [hunk 21]
```diff
@@ -25,7 +25,7 @@ <h1>
                             <img ng-src="{{notification.from.photo_url}}" class="md-avatar" style="width: 60px; height: 60px;"/>
                             <div class="md-list-item-text" layout="column">
                                 <p><b style="font-size: 18px;">{{ notification.from.name }}</b></p>
-                                <p><b style="font-size: 15px; color: #009688;">{{ notification.current_institution.name }}</b></p>
+                                <p><b style="font-size: 15px; color: #009688;">{{ notification.current_institution.name || notification.from.institution_name }}</b></p>
```
Comment (mayzabeel) [inline frontend/notification/notifications_page.html]: insert a comment explaining what is new notification data and what is old notification data

File: backend/service_messages.py [hunk 22]
```diff
@@ -66,7 +66,7 @@ def send_message_notification(receiver_key, entity_type, entity_key, message, en
         params={
             'receiver_key': receiver_key,
             'message': message,
-            'entity_type': entity_type,
+            'entity_type': notification_type,
```
Comment (ruanres) [inline backend/service_messages.py]: faltou esse

**Pull request merged**